### PR TITLE
Add MEDIA_URL and MEDIA_ROOT option

### PR DESCRIPTION
**Pull request by gagelarsen**
Changes Made:
 - Add media to portal config
 - Add media to urls in apps
 - Add media to nginx config

### CODE REVIEW

File: docker/salt/tethyscore.sls [hunk 1]
```diff
@@ -28,6 +28,8 @@
 {% set TETHYS_PERSIST = salt['environ.get']('TETHYS_PERSIST') %}
 {% set STATIC_ROOT = salt['environ.get']('STATIC_ROOT') %}
 {% set WORKSPACE_ROOT = salt['environ.get']('WORKSPACE_ROOT') %}
+{% set WORKSPACE_URL = salt['environ.get']('WORKSPACE_URL') %}
```
Comment (sdc50) [inline docker/salt/tethyscore.sls:31]: Is this supposed to be:

```suggestion
{% set MEDIA_URL = salt['environ.get']('MEDIA_URL') %}
```

Comment (gagelarsen) [reply docker/salt/tethyscore.sls:31]: Yes, thanks for catching. Fixed.

File: Dockerfile [hunk 2]
```diff
@@ -45,6 +45,7 @@ ENV SESSION_WARN=1500
 ENV SESSION_EXPIRE=1800
 ENV STATIC_ROOT="${TETHYS_PERSIST}/static"
 ENV WORKSPACE_ROOT="${TETHYS_PERSIST}/workspaces"
+ENV MEDIA_ROOT="${WORKSPACE_ROOT}"
 ENV QUOTA_HANDLERS="\"[]\""
```
Comment (sdc50) [inline Dockerfile:49]: Do we also need to add a `MEDIA_URL` env var to the `Dockerfile`?

File: Dockerfile [hunk 3]
```diff
@@ -45,6 +45,7 @@ ENV SESSION_WARN=1500
 ENV SESSION_EXPIRE=1800
 ENV STATIC_ROOT="${TETHYS_PERSIST}/static"
 ENV WORKSPACE_ROOT="${TETHYS_PERSIST}/workspaces"
+ENV MEDIA_ROOT="${WORKSPACE_ROOT}"
```
Comment (sdc50) [inline Dockerfile:48]:
```suggestion
ENV MEDIA_ROOT="${WORKSPACE_ROOT}"
ENV MEDIA_URL="/media/"
```

**Comment by sdc50** [comment]
We should create a `media` directory in the app scaffold similar to the `public` directory, but that is runtime writable. 

Create an API to write to the media directory (similar to the `workspaces` API).

**Comment by swainn** [comment]
> We should create a `media` directory in the app scaffold similar to the `public` directory, but that is runtime writable.
> 
> Create an API to write to the media directory (similar to the `workspaces` API).

Except different than workspaces. Don't do collectworkspaces or symlink in production. The API would just know to write to MEDIA_ROOT.

File: tethys_portal/settings.py [hunk 4]
```diff
@@ -452,6 +452,9 @@
     "TETHYS_WORKSPACES_ROOT", relative_to_tethys_home("workspaces", as_str=True)
 )
 
+MEDIA_URL = portal_config_settings.pop("MEDIA_URL", "/media/")
+MEDIA_ROOT = portal_config_settings.pop("MEDIA_ROOT", TETHYS_WORKSPACES_ROOT)
```
Comment (sdc50) [inline tethys_portal/settings.py:456]:
```suggestion
MEDIA_ROOT = portal_config_settings.pop("MEDIA_ROOT", relative_to_tethys_home("media", as_str=True))
```

Should we just have a dedicated default `media` directory? Or does it make sense to have it default to `TETHYS_WORKSPACES_ROOT`?

Comment (sdc50) [reply tethys_portal/settings.py:456]: If we do have a separate `media` directory, should we also have a `media` quota?

File: Dockerfile (hunk 3 above)
Comment (swainn) [reply Dockerfile:48]: Need to create this directory somewhere in the Dockerfile.

Comment (gagelarsen) [reply Dockerfile]: Changed it to {TETHYS_PERSIST}/media" and added it to the MKDIR command

File: Dockerfile (hunk 2 above)
Comment (gagelarsen) [reply Dockerfile:50]: Yes, I think we do. I have added that.

File: tethys_portal/settings.py (hunk 4 above)
Comment (gagelarsen) [reply tethys_portal/settings.py]: from our meeting on friday I think we decided @sdc50 was going to handle this.

File: docs/installation/production/docker/tethys_docker_reference.rst [hunk 5]
```diff
@@ -127,7 +127,12 @@ These environment variables point to important paths in the container. Changing
 |                           | :file:`portal_config.yml`. Defaults to "${TETHYS_PERSIST}/static"                        |
 +---------------------------+------------------------------------------------------------------------------------------+
 | WORKSPACE_ROOT            | Path to the tethys workspaces root folder. This also sets the associated setting in the  |
-|                           | :file:`portal_config.yml`. Defaults to "${TETHYS_PERSIST}/workspaces"                    |
+|                           | :file:`portal_config.yml`. Defaults to "/media/"                                         |
```
Comment (swainn) [inline docs/installation/production/docker/tethys_docker_reference.rst:130]: I'm looking at this on my phone, so it's hard to tell, but I think you modified the wrong line here. This should be on MEDIA_ROOT.

File: docs/installation/production/docker/tethys_docker_reference.rst [hunk 6]
```diff
@@ -127,7 +127,12 @@ These environment variables point to important paths in the container. Changing
...
++---------------------------+------------------------------------------------------------------------------------------+
+| MEDIA_ROOT                | URL to be used fort ethys media. This also sets the associated setting in the            |
```
Comment (swainn) [inline docs/installation/production/docker/tethys_docker_reference.rst:132]: "for tethys media"

File: docs/installation/production/docker/tethys_docker_reference.rst [hunk 7]
```diff
@@ -127,7 +127,12 @@ These environment variables point to important paths in the container. Changing
...
+|                           | :file:`portal_config.yml`. Defaults to ""                         |
++---------------------------+------------------------------------------------------------------------------------------+
+| MEDIA_URL                 | Path to the tethys media root folder. This also sets the associated setting in the       |
```
Comment (swainn) [inline docs/installation/production/docker/tethys_docker_reference.rst:135]: It also seems that the MEDIA_URL and MEDIA_ROOT descriptions are swapped.

File: docs/tethys_portal/configuration.rst [hunk 8]
```diff
@@ -83,6 +83,8 @@ ENABLE_OPEN_PORTAL                                 no login required for Tethys
...
 TETHYS_WORKSPACES_ROOT                             location to which app workspaces will be synced when ``tethys manage collectworkspaces`` is executed. Gathering all workspaces to one location is recommended for production deployments to allow for easier updating and backing up of app data. Defaults to :file:`<TETHYS_HOME>/workspaces`.
 STATIC_ROOT                                        the Django `STATIC_ROOT <https://docs.djangoproject.com/en/3.2/ref/settings/#static-root>`_ setting. Defaults to :file:`<TETHYS_HOME>/static`.
+MEDIA_URL                                          the Django `MEDIA_URL <https://docs.djangoproject.com/en/3.2/ref/settings/#media-url>`_ setting. Defaults to ``'/media/'``.
+MEDIA_ROOT                                         the Django `MEDIA_ROOT <https://docs.djangoproject.com/en/3.2/ref/settings/#media-root>`_ setting. Defaults to :file:`<TETHYS_WORKSPACES_ROOT>`.
```
Comment (swainn) [inline docs/tethys_portal/configuration.rst:87]: Defaults to '~/.tethys/media/'.

File: tests/unit_tests/test_tethys_cli/test_gen_commands.py [hunk 9]
```diff
@@ -73,13 +73,13 @@ def test_generate_command_apache_option(
...
         mock_args.directory = None
         mock_os_path_isfile.return_value = False
-        mock_settings.side_effect = ["/foo/workspace", "/foo/static"]
+        mock_settings.side_effect = ["/foo/workspace", "/foo/static", "/foo/media"]
 
         generate_command(args=mock_args)
 
         mock_os_path_isfile.assert_called_once()
         mock_file.assert_called()
-        mock_settings.assert_called_with("STATIC_ROOT")
```
Comment (swainn) [inline tests/unit_tests/test_tethys_cli/test_gen_commands.py:82]: Is there a reason you're dropping the check for STATIC_ROOT instead of doing both?

**Review by swainn — COMMENTED** [review]
See comments

File: tests/unit_tests/test_tethys_cli/test_gen_commands.py (hunk 9 above)
Comment (gagelarsen) [reply tests/unit_tests/test_tethys_cli/test_gen_commands.py:82]: whoops, added it back in.

File: docs/tethys_portal/configuration.rst (hunk 8 above)
Comment (gagelarsen) [reply docs/tethys_portal/configuration.rst]: fixed

Comment (gagelarsen) [reply docs/tethys_portal/configuration.rst]: fixed

Comment (gagelarsen) [reply docs/tethys_portal/configuration.rst]: fixed

File: docs/installation/production/docker/tethys_docker_reference.rst (hunk 7 above)
Comment (gagelarsen) [reply docs/installation/production/docker/tethys_docker_reference.rst]: fixed

File: docs/installation/production/docker/tethys_docker_reference.rst (hunk 6 above)
Comment (gagelarsen) [reply docs/installation/production/docker/tethys_docker_reference.rst]: fixed

File: docs/installation/production/docker/tethys_docker_reference.rst (hunk 5 above)
Comment (gagelarsen) [reply docs/installation/production/docker/tethys_docker_reference.rst]: fixed

**Review by swainn — APPROVED** [review]

**Pull request merged**